bridges/snowbridge/parachain/primitives/core/src/outbound.rs

```diff
@@ -0,0 +1,408 @@
+		/// Transfer ERC20 tokens
+		TransferToken {
+			/// Address of the ERC20 token
+			token: H160,
+			/// The recipient of the tokens
+			recipient: H160,
+			/// The amount of tokens to transfer
+			amount: u128,
+		},
+	}
+
+	impl AgentExecuteCommand {
+		fn index(&self) -> u8 {
+			match self {
+				AgentExecuteCommand::TransferToken { .. } => 0,
+			}
+		}
+
+		/// ABI-encode the sub-command
+		pub fn abi_encode(&self) -> Vec<u8> {
+			match self {
+				AgentExecuteCommand::TransferToken { token, recipient, amount } =>
+					ethabi::encode(&[
+						Token::Uint(self.index().into()),
+						Token::Bytes(ethabi::encode(&[
+							Token::Address(*token),
+							Token::Address(*recipient),
+							Token::Uint(U256::from(*amount)),
+						])),
+					]),
+			}
+		}
+	}
+
+	/// Message which is awaiting processing in the MessageQueue pallet
+	#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	#[cfg_attr(feature = "std", derive(PartialEq))]
+	pub struct QueuedMessage {
+		/// Message ID
+		pub id: H256,
+		/// Channel ID
+		pub channel_id: ChannelId,
+		/// Command to execute in the Gateway contract
+		pub command: Command,
+	}
+}
+
+#[cfg_attr(feature = "std", derive(PartialEq, Debug))]
+/// Fee for delivering message
+pub struct Fee<Balance>
+where
+	Balance: BaseArithmetic + Unsigned + Copy,
+{
+	/// Fee to cover cost of processing the message locally
+	pub local: Balance,
+	/// Fee to cover cost processing the message remotely
+	pub remote: Balance,
+}
+
+impl<Balance> Fee<Balance>
+where
+	Balance: BaseArithmetic + Unsigned + Copy,
+{
+	pub fn total(&self) -> Balance {
+		self.local.saturating_add(self.remote)
+	}
+}
+
+impl<Balance> From<(Balance, Balance)> for Fee<Balance>
+where
+	Balance: BaseArithmetic + Unsigned + Copy,
+{
+	fn from((local, remote): (Balance, Balance)) -> Self {
+		Self { local, remote }
+	}
+}
+
+/// A trait for sending messages to Ethereum
+pub trait SendMessage: SendMessageFeeProvider {
+	type Ticket: Clone + Encode + Decode;
+
+	/// Validate an outbound message and return a tuple:
+	/// 1. Ticket for submitting the message
+	/// 2. Delivery fee
+	fn validate(
+		message: &Message,
+	) -> Result<(Self::Ticket, Fee<<Self as SendMessageFeeProvider>::Balance>), SendError>;
+
+	/// Submit the message ticket for eventual delivery to Ethereum
+	fn deliver(ticket: Self::Ticket) -> Result<H256, SendError>;
+}
+
+pub trait Ticket: Encode + Decode + Clone {
+	fn message_id(&self) -> H256;
+}
+
+/// A trait for getting the local costs associated with sending a message.
+pub trait SendMessageFeeProvider {
+	type Balance: BaseArithmetic + Unsigned + Copy;
+
+	/// The local component of the message processing fees in native currency
+	fn local_fee() -> Self::Balance;
+}
+
+/// Reasons why sending to Ethereum could not be initiated
+#[derive(Copy, Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, PalletError, TypeInfo)]
+pub enum SendError {
+	/// Message is too large to be safely executed on Ethereum
+	MessageTooLarge,
```

Comment:
I guess description is because this error is, among other causes, thrown if `payload.len() > T::MaxMessagePayloadSize::get()`, i.e. over 2048 bytes on `bridge-hub-rococo` / in production. What sets this limit on Ethereum?

Reply:
In both directions of the bridge, the messages have a fixed upper bound because the message formats as fixed. For example, from Ethereum -> Polkadot, the two messages types are registerToken and sendToken. Both have clear message size upper bounds. From Polkadot -> Ethereum, the same is true. So this `MessageTooLarge` is just an extra check.

When we add Transact (with arbitrary message content and lengths) we will have to revisit the message size check.